File: android/app/src/main/java/com/kreativekoala/scribeai/data/models/Meeting.kt
package com.kreativekoala.scribeai.data.models

import com.google.gson.annotations.SerializedName

/**
 * Meeting model representing a meeting bot session
 */
data class Meeting(
    val id: String,
    @SerializedName("user_id") val userId: String,
    val title: String?,
    @SerializedName("meeting_url") val meetingUrl: String,
    val platform: MeetingPlatform,
    val status: MeetingStatus,
    @SerializedName("scheduled_start") val scheduledStart: String?,
    @SerializedName("actual_start") val actualStart: String?,
    @SerializedName("actual_end") val actualEnd: String?,
    @SerializedName("duration_seconds") val durationSeconds: Int?,
    @SerializedName("error_message") val errorMessage: String?,
    @SerializedName("note_id") val noteId: String?,
    @SerializedName("created_at") val createdAt: String,
    @SerializedName("updated_at") val updatedAt: String,
    @SerializedName("bot_runs") val botRuns: List<BotRun>? = null,
    @SerializedName("meeting_recordings") val meetingRecordings: List<MeetingRecording>? = null
) {
    /**
     * Format duration as mm:ss or hh:mm:ss
     */
    val formattedDuration: String?
        get() {
            val seconds = durationSeconds ?: return null
            val minutes = seconds / 60
            val remainingSeconds = seconds % 60
            return if (minutes > 60) {
                val hours = minutes / 60
                val remainingMinutes = minutes % 60
                String.format("%d:%02d:%02d", hours, remainingMinutes, remainingSeconds)
            } else {
                String.format("%d:%02d", minutes, remainingSeconds)
            }
        }

    /**
     * Whether the meeting is currently active
     */
    val isActive: Boolean
        get() = status.isActive

    /**
     * Whether the meeting can be cancelled
     */
    val isCancellable: Boolean
        get() = status.isCancellable
}

/**
 * Meeting status enum
 */
enum class MeetingStatus {
    @SerializedName("pending") PENDING,
    @SerializedName("bot_joining") BOT_JOINING,
    @SerializedName("in_progress") IN_PROGRESS,
    @SerializedName("recording") RECORDING,
    @SerializedName("processing") PROCESSING,
    @SerializedName("transcribing") TRANSCRIBING,
    @SerializedName("completed") COMPLETED,
    @SerializedName("failed") FAILED,
    @SerializedName("cancelled") CANCELLED;

    val displayText: String
        get() = when (this) {
            PENDING -> "Waiting to join"
            BOT_JOINING -> "Bot joining..."
            IN_PROGRESS -> "In meeting"
            RECORDING -> "Recording"
            PROCESSING -> "Processing"
            TRANSCRIBING -> "Transcribing"
            COMPLETED -> "Completed"
            FAILED -> "Failed"
            CANCELLED -> "Cancelled"
        }

    val isActive: Boolean
        get() = this in listOf(PENDING, BOT_JOINING, IN_PROGRESS, RECORDING, PROCESSING, TRANSCRIBING)

    val isCancellable: Boolean
        get() = this in listOf(PENDING, BOT_JOINING, IN_PROGRESS, RECORDING)
}

/**
 * Meeting platform enum
 */
enum class MeetingPlatform {
    @SerializedName("zoom") ZOOM,
    @SerializedName("google_meet") GOOGLE_MEET,
    @SerializedName("teams") TEAMS,
    @SerializedName("webex") WEBEX,
    @SerializedName("other") OTHER;

    val displayName: String
        get() = when (this) {
            ZOOM -> "Zoom"
            GOOGLE_MEET -> "Google Meet"
            TEAMS -> "Microsoft Teams"
            WEBEX -> "Webex"
            OTHER -> "Other"
        }
}

/**
 * Bot run model
 */
data class BotRun(
    val id: String,
    @SerializedName("meeting_id") val meetingId: String,
    @SerializedName("recall_bot_id") val recallBotId: String?,
    val status: String,
    @SerializedName("join_time") val joinTime: String?,
    @SerializedName("leave_time") val leaveTime: String?,
    @SerializedName("recording_path") val recordingPath: String?,
    @SerializedName("transcript_raw") val transcriptRaw: String?
)

/**
 * Meeting recording model
 */
data class MeetingRecording(
    val id: String,
    @SerializedName("meeting_id") val meetingId: String,
    @SerializedName("bot_run_id") val botRunId: String?,
    @SerializedName("storage_path") val storagePath: String,
    @SerializedName("storage_url") val storageUrl: String?,
    @SerializedName("file_size_bytes") val fileSizeBytes: Long?,
    @SerializedName("duration_seconds") val durationSeconds: Int?,
    val format: String?,
    val status: String,
    @SerializedName("created_at") val createdAt: String
)

// MARK: - API Response Types

data class MeetingResponse(
    val success: Boolean,
    val data: Meeting?,
    val error: String?
)

data class MeetingsListResponse(
    val success: Boolean,
    val data: List<Meeting>?,
    val pagination: MeetingsPagination?,
    val error: String?
)

data class MeetingsPagination(
    val page: Int,
    val limit: Int,
    val total: Int,
    val pages: Int
) {
    val hasMore: Boolean
        get() = page < pages
}

data class CreateMeetingRequest(
    val meetingUrl: String,
    val title: String? = null
)

data class CreateMeetingResponse(
    val success: Boolean,
    val data: CreateMeetingData?,
    val error: String?
)

data class CreateMeetingData(
    val meeting: Meeting,
    val botRun: BotRun?,
    val recallBotId: String?
)

data class ValidateMeetingUrlRequest(
    val meetingUrl: String
)

data class ValidateMeetingUrlResponse(
    val success: Boolean,
    val data: ValidateMeetingUrlData?,
    val error: String?
)

data class ValidateMeetingUrlData(
    val valid: Boolean,
    val platform: MeetingPlatform?
)

// MARK: - URL Validation Helper

object MeetingURLValidator {
    /**
     * Validate a meeting URL and detect the platform
     */
    fun validate(url: String): Pair<Boolean, MeetingPlatform?> {
        val lowercased = url.lowercase()

        return when {
            lowercased.contains("zoom.us/j/") || lowercased.contains("zoom.us/my/") ->
                true to MeetingPlatform.ZOOM
            lowercased.contains("meet.google.com/") ->
                true to MeetingPlatform.GOOGLE_MEET
            lowercased.contains("teams.microsoft.com") || lowercased.contains("teams.live.com") ->
                true to MeetingPlatform.TEAMS
            lowercased.contains("webex.com") ->
                true to MeetingPlatform.WEBEX
            else -> false to null
        }
    }
}

File: ios/scribeai/Models/Meeting.swift
//
//  Meeting.swift
//  scribeai
//
//  Meeting models for meeting bot functionality
//

import Foundation

// MARK: - Meeting Model

struct Meeting: Identifiable, Codable {
    let id: String
    let userId: String
    let title: String?
    let meetingUrl: String
    let platform: MeetingPlatform
    let status: MeetingStatus
    let scheduledStart: String?
    let actualStart: String?
    let actualEnd: String?
    let durationSeconds: Int?
    let errorMessage: String?
    let noteId: String?
    let createdAt: String
    let updatedAt: String

    var botRuns: [BotRun]?
    var meetingRecordings: [MeetingRecording]?

    enum CodingKeys: String, CodingKey {
        case id, title, platform, status
        case userId = "user_id"
        case meetingUrl = "meeting_url"
        case scheduledStart = "scheduled_start"
        case actualStart = "actual_start"
        case actualEnd = "actual_end"
        case durationSeconds = "duration_seconds"
        case errorMessage = "error_message"
        case noteId = "note_id"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case botRuns = "bot_runs"
        case meetingRecordings = "meeting_recordings"
    }

    /// Formatted duration string
    var formattedDuration: String? {
        guard let seconds = durationSeconds else { return nil }
        let minutes = seconds / 60
        let remainingSeconds = seconds % 60
        if minutes > 60 {
            let hours = minutes / 60
            let remainingMinutes = minutes % 60
            return String(format: "%d:%02d:%02d", hours, remainingMinutes, remainingSeconds)
        }
        return String(format: "%d:%02d", minutes, remainingSeconds)
    }

    /// Whether the meeting is currently in progress
    var isActive: Bool {
        status.isActive
    }

    /// Whether the meeting can be cancelled
    var isCancellable: Bool {
        status.isCancellable
    }
}

// MARK: - Meeting Status

enum MeetingStatus: String, Codable, CaseIterable {
    case pending
    case botJoining = "bot_joining"
    case inProgress = "in_progress"
    case recording
    case processing
    case transcribing
    case completed
    case failed
    case cancelled

    var displayText: String {
        switch self {
        case .pending: return "Waiting to join"
        case .botJoining: return "Bot joining..."
        case .inProgress: return "In meeting"
        case .recording: return "Recording"
        case .processing: return "Processing"
        case .transcribing: return "Transcribing"
        case .completed: return "Completed"
        case .failed: return "Failed"
        case .cancelled: return "Cancelled"
        }
    }

    var isActive: Bool {
        switch self {
        case .pending, .botJoining, .inProgress, .recording, .processing, .transcribing:
            return true
        case .completed, .failed, .cancelled:
            return false
        }
    }

    var isCancellable: Bool {
        switch self {
        case .pending, .botJoining, .inProgress, .recording:
            return true
        default:
            return false
        }
    }

    var iconName: String {
        switch self {
        case .pending: return "clock"
        case .botJoining: return "arrow.right.circle"
        case .inProgress: return "person.3.fill"
        case .recording: return "record.circle"
        case .processing: return "gearshape.2"
        case .transcribing: return "text.bubble"
        case .completed: return "checkmark.circle.fill"
        case .failed: return "xmark.circle.fill"
        case .cancelled: return "slash.circle"
        }
    }
}

// MARK: - Meeting Platform

enum MeetingPlatform: String, Codable, CaseIterable {
    case zoom
    case googleMeet = "google_meet"
    case teams
    case webex
    case other

    var displayName: String {
        switch self {
        case .zoom: return "Zoom"
        case .googleMeet: return "Google Meet"
        case .teams: return "Microsoft Teams"
        case .webex: return "Webex"
        case .other: return "Other"
        }
    }

    var iconName: String {
        // All use video icon but could be customized per platform
        return "video.fill"
    }

    var color: String {
        switch self {
        case .zoom: return "blue"
        case .googleMeet: return "green"
        case .teams: return "purple"
        case .webex: return "orange"
        case .other: return "gray"
        }
    }
}

// MARK: - Bot Run

struct BotRun: Identifiable, Codable {
    let id: String
    let meetingId: String
    let recallBotId: String?
    let status: String
    let joinTime: String?
    let leaveTime: String?
    let recordingPath: String?
    let transcriptRaw: String?

    enum CodingKeys: String, CodingKey {
        case id, status
        case meetingId = "meeting_id"
        case recallBotId = "recall_bot_id"
        case joinTime = "join_time"
        case leaveTime = "leave_time"
        case recordingPath = "recording_path"
        case transcriptRaw = "transcript_raw"
    }
}

// MARK: - Meeting Recording

struct MeetingRecording: Identifiable, Codable {
    let id: String
    let meetingId: String
    let botRunId: String?
    let storagePath: String
    let storageUrl: String?
    let fileSizeBytes: Int?
    let durationSeconds: Int?
    let format: String?
    let status: String
    let createdAt: String

    enum CodingKeys: String, CodingKey {
        case id, status, format
        case meetingId = "meeting_id"
        case botRunId = "bot_run_id"
        case storagePath = "storage_path"
        case storageUrl = "storage_url"
        case fileSizeBytes = "file_size_bytes"
        case durationSeconds = "duration_seconds"
        case createdAt = "created_at"
    }
}

// MARK: - API Response Types

struct MeetingResponse: Codable {
    let success: Bool
    let data: Meeting?
    let error: String?
}

struct MeetingsListResponse: Codable {
    let success: Bool
    let data: [Meeting]?
    let pagination: MeetingsPagination?
    let error: String?
}

struct MeetingsPagination: Codable {
    let page: Int
    let limit: Int
    let total: Int
    let pages: Int

    var hasMore: Bool {
        page < pages
    }
}

struct CreateMeetingResponse: Codable {
    let success: Bool
    let data: CreateMeetingData?
    let error: String?
}

struct CreateMeetingData: Codable {
    let meeting: Meeting
    let botRun: BotRun?
    let recallBotId: String?
}

struct ValidateMeetingUrlResponse: Codable {
    let success: Bool
    let data: ValidateMeetingUrlData?
    let error: String?
}

struct ValidateMeetingUrlData: Codable {
    let valid: Bool
    let platform: MeetingPlatform?
}

struct SupportedPlatformsResponse: Codable {
    let success: Bool
    let data: [SupportedPlatform]?
    let error: String?
}

struct SupportedPlatform: Codable {
    let id: String
    let name: String
    let urlExample: String
}

// MARK: - URL Validation Helper

struct MeetingURLValidator {
    /// Validate a meeting URL and detect the platform
    static func validate(_ url: String) -> (valid: Bool, platform: MeetingPlatform?) {
        let lowercased = url.lowercased()

        if lowercased.contains("zoom.us/j/") || lowercased.contains("zoom.us/my/") {
            return (true, .zoom)
        }

        if lowercased.contains("meet.google.com/") {
            return (true, .googleMeet)
        }

        if lowercased.contains("teams.microsoft.com") || lowercased.contains("teams.live.com") {
            return (true, .teams)
        }

        if lowercased.contains("webex.com") {
            return (true, .webex)
        }

        // Check for valid URL format but unknown platform
        if let urlObj = URL(string: url), urlObj.scheme?.hasPrefix("http") == true {
            return (false, nil)
        }

        return (false, nil)
    }
}
